### PR TITLE
Add support for custom backend

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95072

Fixes https://github.com/pytorch/pytorch/issues/92344

A custom backend can be specified by passing in a string with format `"<device_type1>:<backend_name>,<device_type2>:<backend_name>"`, e.g. `"cpu:gloo,cuda:custom_backend"`.

Differential Revision: [D43630050](https://our.internmc.facebook.com/intern/diff/D43630050)

cc @ezyang @gchanan @zou3519

### CODE REVIEW

**Comment by Zha0q1** [comment]
Thanks @H-Huang for starting this PR!

> A custom backend can be specified by passing ina string with format "<device_type1>:<backend_name>,<device_type2>:<backend_name>", e.g. "cpu:gloo,cuda:custom_backend"
Will dist.init_process_group(backend='custom') work? If so, will all devices be mapped to "custom"? Asking because existing frameworks and training scripts have been using this semantic.

**Comment by H-Huang** [comment]
Hi @Zha0q1,

> Will dist.init_process_group(backend='custom') work? If so, will all devices be mapped to "custom"? Asking because existing frameworks and training scripts have been using this semantic.

Yes, that will continue to work as it does in PT1.0. `dist.init_process_group(backend='custom')` will be equivalent to config `cpu:custom,cuda:custom`

**Comment by Zha0q1** [comment]
Hi @H-Huang, we will start testing this PR from our side too, thank you so much for making this happen! Just wanted to confirm, is this PR pending any additional commits, or we can already take it to testing?

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43630050).

**Comment by H-Huang** [comment]
@Zha0q1 you can test this PR! It is under review, but I don't expect any major changes

**Comment by Zha0q1** [comment]
I have built and tested the latest commit ede42e6a24b95dca7643c8ef8ec053808b2d3b93
Custom backend is correctly handled now -- 
`backend='custom'` creates only one backend to handle both cpu and cuda tensors
`backend=cpu:gloo,cuda:custom` creates a gloo backend for cpu and a custom backend for cuda

File: torch/distributed/distributed_c10d.py
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:266]: I wonder if this check would be too restrictive for non-CUDA, non-CPU use cases?

File: torch/distributed/distributed_c10d.py
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:1108]: Maybe:
`if len(...) == 1`
to make it simpler?

File: torch/distributed/distributed_c10d.py
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:248]: Maybe more comments for this syntax if you can?

An idea for avoiding using `re` at all: 
just go ahead to `split` the string by `,` and then split by `:` (as your code below does). If the returned split list's length `!= 2`, that means either `:` is not found or there are too many `:`.

File: torch/distributed/distributed_c10d.py
Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:1108]: Thanks! updated

File: torch/distributed/distributed_c10d.py
Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:266]: Hmm yeah, I suppose that if user doesn't specify "cpu" or "cuda" then process group won't error out unless explicitly called with a "cpu" or "cuda" tensor, so we don't need this. I can remove this check

File: torch/distributed/distributed_c10d.py
Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:248]: Yeah I agree the check can be done that way. I think regex might be a little more manageable and result in less if conditions. It also limits the inputs to alphanumeric characters. For example if user does something accidentally like `cpu;:gloo` then it would error with regex (due to the `;`) but we would need another case to check for special characters. If you want to checks for each `:` pair I can add that instead, lmk!

Updated the comments as well.

Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:250]: > It also limits the inputs to alphanumeric characters

Then something like "my_awesome_backend" would not work?

Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:250]: hmm thats right, do you think it is necessary?

**Review by kwen2501 — APPROVED** [review]
Thanks for the update. LGTM!

**Comment by atalman** [comment]
cc @H-Huang 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Pull request closed without merging**
